### PR TITLE
Add dynamic badges functionality

### PR DESCRIPTION
**Pull request by nikhilrayaprolu**

### CODE REVIEW

**Comment by nikhilrayaprolu** [comment]
Dynamic Badges
image

custom fields in Badges
image

**Comment by nikhilrayaprolu** [comment]
image

**Comment by nikhilrayaprolu** [comment]
Linkes to the pages (Login to view the pages)
https://aicrowd-feature-dynamic-ehcqa6.herokuapp.com/aicrowd_badges
https://aicrowd-feature-dynamic-ehcqa6.herokuapp.com/badges_events

File: app/controllers/badges_events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class BadgesEventsController < InheritedResources::Base
```
Comment (piotrekpasciak) [inline app/controllers/badges_events_controller.rb:1]: Please move it to ActiveAdmin

File: app/javascript/stimulus_js/controllers/aicrowd_badges_controller.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { Controller } from 'stimulus';
+
+export default class extends Controller {
+
```
Comment (piotrekpasciak) [inline app/javascript/stimulus_js/controllers/aicrowd_badges_controller.js:4]: Please fix the identation.

File: app/javascript/stimulus_js/controllers/aicrowd_badges_controller.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    connect(){
+        var editor = ace.edit("code-block");
```
Comment (piotrekpasciak) [inline app/javascript/stimulus_js/controllers/aicrowd_badges_controller.js:6]: Use `const` instead of `var` for javasccript variables.

File: app/jobs/badges_daily_job.rb [hunk 4]
```diff
@@ -38,12 +38,7 @@ def assign_badges_to_challenge(round_id)
...
     max_rank = ranks.last
-    during_badge = 0
-    if user_rating_service.temporary
-      during_badge = 3
-    else
-      apply_badges(teams_participant_ids, during_badge, ranks, max_rank)
-      user_rating_service.round.update(assigned_permanent_badge: true)
-    end
+    apply_badges(teams_participant_ids, ranks, max_rank, round_id)
+    user_rating_service.round.update(assigned_permanent_badge: true)
```
Comment (piotrekpasciak) [inline app/jobs/badges_daily_job.rb:42]: `update!`

File: app/models/participant.rb [hunk 5]
```diff
@@ -149,6 +149,19 @@ def datetime_sequence(start, stop, step)
     return dates
   end
 
+  def add_badge(name, custom_fields={})
+    badge_id = AicrowdBadge.find_by(name: name).id
+    AicrowdUserBadge.create(aicrowd_badge_id: badge_id, participant_id: self.id, custom_fields: custom_fields)
```
Comment (piotrekpasciak) [inline app/models/participant.rb:154]: `create!`

Comment (piotrekpasciak) [reply app/models/participant.rb:154]: Please don't use `self`. when it's not needed:
https://github.com/rubocop-hq/ruby-style-guide#explicit-self

File: app/models/participant.rb [hunk 6]
```diff
@@ -149,6 +149,19 @@ def datetime_sequence(start, stop, step)
...
+  end
+
+  def rm_badge(name)
+    badge_id = AicrowdBadge.find_by(name: name).id
+    AicrowdUserBadge.find_by(aicrowd_badges_id: badge_id, participant_id: self.id).destroy
+  end
+
+  def badges
```
Comment (piotrekpasciak) [inline app/models/participant.rb:162]: Why do you need this method? What's the difference between this method and
`has_many :aicrowd_user_badges` ?

File: db/schema.rb [hunk 7]
```diff
@@ -434,6 +472,7 @@
     t.integer "team_freeze_seconds_before_end", default: 604800
     t.boolean "hidden_challenge", default: false, null: false
     t.datetime "team_freeze_time"
+    t.string "evaluator_type_cd"
```
Comment (piotrekpasciak) [inline db/schema.rb:475]: This doesn't look like part of this PR. Please remove it.

File: db/migrate/20200501130736_add_index_to_aicrowd_badges.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToAicrowdBadges < ActiveRecord::Migration[5.2]
+  def change
+    add_index :aicrowd_badges, :name, unique: true
```
Comment (piotrekpasciak) [inline db/migrate/20200501130736_add_index_to_aicrowd_badges.rb:3]: Please add corresponding validation to model as well.

File: db/migrate/20200501130845_add_index_to_badges_events.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToBadgesEvents < ActiveRecord::Migration[5.2]
+  def change
+    add_index :badges_events, :name, unique: true
```
Comment (piotrekpasciak) [inline db/migrate/20200501130845_add_index_to_badges_events.rb:3]: Please add ActiveRecord validation to model.

File: db/migrate/20200427171543_create_badges_events.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class CreateBadgesEvents < ActiveRecord::Migration[5.2]
+  def change
+    create_table :badges_events do |t|
+      t.string :name
```
Comment (piotrekpasciak) [inline db/migrate/20200427171543_create_badges_events.rb:4]: This should be `null: false` and model should validate presence.

File: db/migrate/20200427144932_create_badge_types.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class CreateBadgeTypes < ActiveRecord::Migration[5.2]
+  def change
+    create_table :badge_types do |t|
+      t.string :name
```
Comment (piotrekpasciak) [inline db/migrate/20200427144932_create_badge_types.rb:4]: This should be `null: false` and model should validate presence.

File: db/migrate/20200427144206_create_aicrowd_badges.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class CreateAicrowdBadges < ActiveRecord::Migration[5.2]
+  def change
+    create_table :aicrowd_badges do |t|
+      t.string :name
```
Comment (piotrekpasciak) [inline db/migrate/20200427144206_create_aicrowd_badges.rb:4]: This should be `null: false` and model should validate presence.

File: app/views/badges_events/index.json.jbuilder [hunk 13]
```diff
@@ -0,0 +1 @@
+json.array! @badges_events, partial: 'badges_events/badges_event', as: :badges_event
```
Comment (piotrekpasciak) [inline app/views/badges_events/index.json.jbuilder:1]: What's the purpose of this file? We already use a different approach to serializing:
https://github.com/AIcrowd/AIcrowd/blob/master/app/serializers/api/v1/api_user_serializer.rb

File: app/models/participant.rb (hunk 6 above)
Comment (nikhilrayaprolu) [reply app/models/participant.rb:162]: is there a helper function that I can use to retrieve the related objects when related with `has_many`
I am not sure, so had to create this method

File: db/migrate/20200501130736_add_index_to_aicrowd_badges.rb (hunk 8 above)
Comment (nikhilrayaprolu) [reply db/migrate/20200501130736_add_index_to_aicrowd_badges.rb:3]: added

File: app/models/participant.rb (hunk 6 above)
Comment (piotrekpasciak) [reply app/models/participant.rb:162]: Screenshot 2020-05-10 at 15 21 49

Please take a look on screenshot above. `aicrowd_user_badges` method works exactly the same as `badges`. Please remove `badges` method.

File: app/admin/aicrowd_badges.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
+ActiveAdmin.register AicrowdBadge do
+# See permitted parameters documentation:
```
Comment (piotrekpasciak) [inline app/admin/aicrowd_badges.rb:2]: Please remove commented code.

File: app/admin/badges_events.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+ActiveAdmin.register BadgesEvent do
+# See permitted parameters documentation:
```
Comment (piotrekpasciak) [inline app/admin/badges_events.rb:2]: Please remove commented code.

File: db/schema.rb [hunk 16]
```diff
@@ -9,7 +9,8 @@
 # you'll amass, the slower it'll run and the greater likelihood for issues).
 #
 # It's strongly recommended that you check this file into your version control system.
-ActiveRecord::Schema.define(version: 2020_04_30_234339) do
```
Comment (piotrekpasciak) [inline db/schema.rb:12]: schema.rb doesn't include changed for dropped tables

File: app/assets/stylesheets/application.css.scss [hunk 17]
```diff
@@ -0,0 +1,165 @@
+// Variables
```
Comment (piotrekpasciak) [inline app/assets/stylesheets/application.css.scss:1]: Why did you add this file? As far I remember I've removed it from repository.

Comment (piotrekpasciak) [inline app/assets/stylesheets/application.css.scss:1]: Why did you readd this file? I remember that I've removed it because it wasn't used in application.
